data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.peerforwarder;
+
+import com.amazon.dataprepper.model.event.Event;
+import com.amazon.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.peerforwarder.discovery.StaticPeerListProvider;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import org.opensearch.dataprepper.peerforwarder.client.PeerForwarderClient;
+
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class PeerForwarder {
+    private static final WebClient LOCAL_CLIENT = null;
+
+    private final PeerForwarderClientFactory peerForwarderClientFactory;
+    private final PeerForwarderClient peerForwarderClient;
+    private HashRing hashRing;
+    private PeerClientPool peerClientPool;
+
+    public PeerForwarder(final PeerForwarderClientFactory peerForwarderClientFactory,
+                         final PeerForwarderClient peerForwarderClient) {
+        this.peerForwarderClientFactory = peerForwarderClientFactory;
+        this.peerForwarderClient = peerForwarderClient;
+    }
+
+    public List<Record<Event>> forwardRecords(final Collection<Record<Event>> records, final Set<String> identificationKeys) {
+        // TODO: decide the default behaviour of Core Peer Forwarder and move the HashRing and PeerClientPool creation to constructor
+        hashRing = peerForwarderClientFactory.createHashRing();
+        peerClientPool = peerForwarderClientFactory.setPeerClientPool();
+
+        final Map<String, List<Record<Event>>> groupedRecords = groupRecordsBasedOnIdentificationKeys(records, identificationKeys);
+
+        final List<Record<Event>> recordsToProcessLocally = new ArrayList<>();
+
+        for (final Map.Entry<String, List<Record<Event>>> entry: groupedRecords.entrySet()) {
+            final WebClient client = getClient(entry.getKey());
```

Comment:
Going off my comment below about not passing in a `WebClient`, this might be refactored like the following.

```
String destinationIp = entry.getKey();
if(isAddressDefinedLocally(destinationIp)) {
  // same as before
} else {
  // same as before
}
```